### ISSUE TITLE
[Bug]: Packaged application with icons for various DPIs inflates memory usage by over 500mb

### ISSUE THREAD

**Issue by Slapbox** [electron/electron#32359]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

13.6.0

### What operating system are you using?

Windows

### Operating System Version

Windows 10

### What arch are you using?

x64

### Last Known Working Electron version

Unknown

### Expected Behavior

Including @1x @2x and so on high DPI image files should not affect memory usage more than ~1mb.

### Actual Behavior

All included variations seem to be loaded into memory even if they're not needed, and the memory allocated for them is over ten times the actual size of the files.

### Testcase Gist URL

N/A - Can't create an Electron Fiddle for an issue that affects packaged applications

### Additional Information

App was packaged with Electron-Builder - but the issue appears to be with Electron itself.

**Comment by ckerr** [comment]
Thanks for reporting this and helping to make Electron better!

Would it be possible for you to make a standalone testcase with only the code necessary to reproduce the issue? [Electron Fiddle](https://www.electronjs.org/fiddle) is a great tool for making small testcases and publishing them as [gists](https://gist.github.com) that Electron maintainers can use.

If the bug you've found can be tested with a pass / fail test, please make the testcase [exit 0 on success or nonzero on failure](https://github.com/electron/bugbot#by-bug-reporters). The lets [Electron's issue bot](https://github.com/electron/bugbot#readme) see what releases are affected by the bug by checking your test against different OSes and Electron versions.

I'm adding the `blocked/need-repro` label for this reason. After you make a test case, please link to it in a followup comment.

Thanks in advance! Your help is appreciated.

**Comment by Slapbox** [comment]
@ckerr I'd be happy to, but how can I if the issue affects only the packaged application?

I am not yet certain it only affects the packaged application - but if it does, then what would be the next step?

**Comment by nornagon** [comment]
@Slapbox a small repository derived from [electron-quick-start](/electron/electron-quick-start) that exhibits the issue would work.

**Comment by Slapbox** [comment]
@nornagon @ckerr no luck reproing this so far but it's definitely a real issue we're seeing. 

The memory accumulations are all in a single process started by Electron. We use a [custom protocol](https://github.com/reZach/secure-electron-template/blob/master/app/electron/protocol.js) in our packaged application, which I haven't tried implementing in our repro case. Could registering a custom protocol be a possible cause?

Is there any repo you know of that demonstrates using icons with multiple DPIs? Do you have any advice on how we might isolate this issue?

I'm at a loss, but obviously 500mb of memory usage is a show stopper for releasing our app with support for multiple DPIs. Thank you for any advice.

**Comment by Slapbox** [comment]
Storing images in a package within a monorepo _may_ be the cause.

**Comment by nornagon** [comment]
It's not really possible for me to comment usefully without more information. If you can reproduce this in an isolated way, I can likely help!

**Comment by Slapbox** [comment]
I understand @nornagon. Thanks for your reply.

It turned out not to be the monorepo issue, and I really have no idea what to investigate that I haven't already.

----

All I can say with certainty is that the `main` process is using an extra half gigabyte of memory unless we include only one DPI for the icon file.

![image](https://user-images.githubusercontent.com/6835891/150023237-37d3a010-a5a5-4c36-be94-98288925cba4.png)

----

Our code to load the images looks like this:


```js
const baseMediaPath = path.join(rootPath, 'assets', 'icons');

const getMediaAsset = (...pathSegments) =>
  path.join(baseMediaPath, ...pathSegments);

trayIcon = nativeImage.createFromPath(getMediaAsset('icon-1024x1024.png'));
```

If anyone has any input on what the cause might be, I'd love to get the bottom of this - but I'm out of ideas for now.

**Comment by nornagon** [comment]
@Slapbox you might try taking a trace with the `memory-infra` category enabled, docs here: https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/memory-infra. There's also some support for [heap profiling](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/memory-infra/heap_profiler.md), though I don't think we expose that yet in Electron.

Also try taking a heap snapshot in devtools (in the main process), on the off chance that can capture it.

**Comment by Slapbox** [comment]
Thank you so much for the tips @nornagon!

A heap snapshot from the `main` process doesn't appear to capture the issue, unfortunately. 

The JS heap size is about the same, but total memory usage in Task Manager is about 795mb higher when multi-DPI images are present.

----

**Regarding `memory-infra`:**

Do you know if there's a minimum Electron version required to enable `memory-infra`? Is this worth pursuing if heap profiling isn't yet exposed?

I reviewed those links and the Electron docs and I think I understand - I would use [`contentTracing`](https://www.electronjs.org/docs/latest/api/content-tracing) with a custom [`TraceConfig`](https://www.electronjs.org/docs/latest/api/structures/trace-config) to enable `memory-infra`, and then load the generated trace in Chrome with `chrome://tracing`?

**So like:**

```js
await contentTracing.startRecording({
  included_categories: ['memory-infra']
})
```

Should there be any other categories enabled?

**Comment by nornagon** [comment]
@Slapbox ah, you need to specify `disabled-by-default-memory-infra`.

Here's an example fiddle: https://gist.github.com/b690aff14017591f95f4ec23a5531b30

**Comment by Slapbox** [comment]
That's a huge help! Thanks @nornagon!

I was able to capture discrepancies in memory usage using the trace results, though I'm not sure whether the results are sufficiently detailed to identify the cause of the issue.

Multi-DPI images result in about 284mb of additional memory usage in `v8` - but the private footprint is almost 900mb higher. 

I'm not sure where the other roughly 568mb of private memory usage might be getting used. About 13mb is accounted for by `cc` (compositor) and `gpu`. That still leaves most of the private memory usage difference unaccounted for though.

![image](https://user-images.githubusercontent.com/6835891/150433598-83c4bee8-df6c-4256-92e1-7efacfff8034.png)

Would it be helpful to share the trace result files themselves? If so I'll anonymize them and share (any advice about how to share those?) Each trace result `json` file is about 38mb.

A big thank you again for your help in investigating this very strange issue!

**Comment by nornagon** [comment]
Looks to me like the stuff is on the v8 heap! That's interesting. You said you didn't manage to capture anything that looked large using devtools, though?

**Comment by Slapbox** [comment]
Yeah it's really strange - the heap size seems large to me (297mb) but it's the same with/without multi-DPI images - whereas what `memory-infra` records differs between the two cases.

I performed each test twice and selected around 11,000ms on the chart and the results seem pretty consistent, so I don't think this is just a case of the garbage collector running or not (especially since I see the issue persistently when the app has been opened and in use long enough that garbage collection should have occurred many times.)

Before capturing the heap snapshot the memory usage is around 650mb in both cases, and then recording the snapshot drops the memory usage down to 297mb, I guess because the garbage collector is run before taking the snapshot.

----

I wonder too, where that additional memory usage in the `private footprint` is coming from. The `v8` memory usage is definitely up, but the private memory footprint is even more confusing to me since it seems unaccounted for.

**Comment by nornagon** [comment]
@Slapbox 'private footprint' is just a sum of all the non-shared memory usage.

**Comment by nornagon** [comment]
I guess I'm back to not really being able to help without a repro :/ presumably it involves more than just creating a nativeImage for each of the images you've linked. Without knowing where those images go in your app, it's hard for me to form a hypothesis.

**Comment by Slapbox** [comment]
The (perhaps) good news is that the image is solely used for the tray icon and window icon; nowhere else at all. That's also what makes the enormous memory usage so totally baffling. On the front end, we use SVGs exclusively, no `nativeImage` usage at all.

So all that memory usage is solely for the tray icon and the window icon that appears in the taskbar/application switcher ¯\\\_(ツ)\_/¯

We assign the `nativeImage` and the generated `Tray` object to global variable slike below. I wonder if that could have some impact somehow? I can't really see how, but I'm grasping at straws.

```js
global.trayIconImage = nativeImage.createFromPath(
  getMediaAsset('shared/icons/icon-1024x1024.png')
);

global.appIcon = new Tray(global.trayIconImage);
```

**Comment by nornagon** [comment]
@Slapbox and doing that in a test app, e.g. Fiddle, doesn't cause the same memory usage situation?

**Comment by Slapbox** [comment]
I was working with the [electron-quick-start](https://github.com/electron/electron-quick-start) template, and unfortunately not able to replicate it so far =/

I'll play with it in a Fiddle, now that I've proved the memory usage is inflated even in development.

I've tried both building up a new repo to produce the problem and tearing down our project to try to remove the problem - but no amount of tearing things away seems to make any difference at all in the memory usage - except removing the high DPI versions from our icons directory.

I guess my next step is to try to lower my Electron version to see if the bug was present with earlier versions (we hadn't yet implemented high DPI icons when we were using earlier versions.)

Drat... I wish the memory usage wasn't so extreme that this was such a major issue for us.

**Comment by nornagon** [comment]
I'm looking forward to finding the answer to this! It's certainly a strange one.

**Comment by Slapbox** [comment]
Thanks so much for all your guidance and support @nornagon!

Until/unless I find the root cause, I'm going to keep adding to my findings here in case anything stands out:

### Testing older Electron versions:
Same issue back to 9.x.x


### Variations of icon usage:

The same source icon files are used for both the tray icon and the window icon, _but_, we create separate `nativeImage`s from that source icon file (one for the tray icon and one for the window icon)

| Case                                        | Memory usage |
|---------------------------------------------|--------------|
| No DPI variations                     | 1354mb       |
| DPI variations only for window icon   | 1683mb       |
| DPI variations only for tray icon     | 1919mb       |
| DPI variations for window & tray icon | 2030mb       |


### Scale factor variations:
Larger scale factors seem to exacerbate the issue more than image count

| Case                               | Memory usage | Size on disk |
|------------------------------------|--------------|--------------|
| DPIs below @2x only (6 images)     | 1682mb       | 151kb        |
| DPIs @2x and above only (5 images) | 1851mb       | 389kb        |

**Comment by Slapbox** [comment]
Linux does not appear to be affected, 1497mb memory usage with or without high DPI icons.

**Comment by Slapbox** [comment]
Finally nailed down the cause. Here's a repro, embarrassingly simple: https://github.com/Slapbox/electron-icon-memory-usage-bug

See the project's README for simple instructions on usage.

TLDR version: Assigning a multi-dpi `nativeImage` to a property of the `global` variable leads to major memory usage inflation on Windows operating systems. I'm confused about why the issue doesn't seem to affect Linux, but happy there's a repro now.

**Comment by nornagon** [comment]
Slightly simplified, + print out memory usage:

```js
const {app, nativeImage} = require('electron')
const path = require('path')
setInterval(() => {
  gc()
  console.log(app.getAppMetrics().find(x => x.type==='Browser'))
}, 1000)

app.whenReady().then(() => {
  const iconPath = path.resolve('build/icons/icon-1024x1024.png');

  //global.appIcon = nativeImage.createFromPath(iconPath); // Main process memory usage on Windows: 312mb
  const appIcon = nativeImage.createFromPath(iconPath); // Main process memory usage on Windows: 18mb
})
```

seems to repro on macOS. Makes sense, having it in the global scope retains the image data, whereas if it's a local variable then it gets GC'd.

**Comment by nornagon** [comment]
I think this is because Skia stores the uncompressed data.

Here are the pixel sizes of your icons:

```
$ identify build/icons/icon-1024x1024@* | cut -d' ' -f1,3
build/icons/icon-1024x1024@1.25x.png 1280x1280
build/icons/icon-1024x1024@1.33x.png 1362x1362
build/icons/icon-1024x1024@1.4x.png 1434x1434
build/icons/icon-1024x1024@1.5x.png 1536x1536
build/icons/icon-1024x1024@1.8x.png 1844x1844
build/icons/icon-1024x1024@1x.png 1024x1024
build/icons/icon-1024x1024@2.5x.png 2560x2560
build/icons/icon-1024x1024@2x.png 2048x2048
build/icons/icon-1024x1024@3x.png 3072x3072
build/icons/icon-1024x1024@4x.png 4096x4096
build/icons/icon-1024x1024@5x.png 5120x5120
```

That's a total of 75534712 pixels. If each pixel is stored as RGBA, 32 bits per pixel, that's a total of about 300 MB, which more or less accounts for the memory usage.

**Comment by Slapbox** [comment]
It sort of makes sense to me, and I can't believe it took me so long to figure this out, but I swore I'd already tested this in the past. COVID brain fog is a big pain.

On the other hand, it sort of doesn't make sense to me:

1. The memory usage is orders of magnitude greater than the image size on disk. I'm guessing this is because `nativeImage` is going to be more akin to a bitmap, yeah?
2. The issue doesn't occur for me on Linux - This is the really baffling thing for me, and part of why the issue went unnoticed by me for many weeks of development
3. The memory used isn't entirely accounted for in `v8`'s memory usage

**Comment by nornagon** [comment]
We do try to report the memory usage to v8: https://github.com/electron/electron/blob/63908ccf89ff8a3f06eee8335675750c12497b86/shell/common/api/electron_api_native_image.cc#L131-L139 perhaps that's not always respected?

**Comment by nornagon** [comment]
This behavior seems to also show on Linux for me. ~400MB w/ the images, ~100MB w/out.

**Comment by Slapbox** [comment]
I just got to test the repro proper on Linux and yeah, result is the same here. But it's not in our actual project, which is why I assumed the same for the repro case. I won't even bother to guess as to why.

So I guess I'm really only still confused about why the memory usage isn't entirely accounted for in `v8` - Not terribly important, and I'm guessing this is something development-environment specific, like how base64 sourcemap strings can inflate memory usage substantially. 

----

**Docs:**
I figured the `global` variable's memory usage would top out at the size of the file that was appropriate for the screen DPI, rather than keeping each variation in memory, regardless of variable scope. I think it would be worth adding a note to the `nativeImage` docs to warn against assignment to `global`s - if you agree I'll make a PR for that the next time I'm working.

**One last question:**
Does that mean someone on a 5x pixel density screen is going to have to devote 80mb of memory to storing an image icon, regardless of variable scope? I don't have any such device to test that (and I don't know of any way to emulate it.) It would be quite brutal if that's the case - am I making a wrong assumption here?

**Comment by nornagon** [comment]
> 5x pixel density screen

does this exist?

**Comment by Slapbox** [comment]
To be honest, I'm not sure. I figured it did because Electron allows @5x images, but a quick look at available laptop screens looks like they may top out at 4x the "standard" pixel density (96ppi.)

But the same question applies for @4x then, which would be around 70mb of memory for the icon alone, if I'm not mistaken, or even @3x would be nearly 38mb of memory usage.

**Comment by Slapbox** [comment]
Unfortunately, after removing all assignments of icons/windows/`Tray` from the `global` object, the excess memory usage is only reduced by 50%, so I guess `global` was not the only issue. I'm again at a loss to explain the discrepancy.

Memory usage is still about 297mb higher when including high DPI variations in the folder.

@nornagon, is there a way to tell what DPI Electron thinks it's supposed to be displaying? Obviously we can't rely on [`window.devicePixelRatio`](https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio) since `window` doesn't exist in the `main` process.

**Comment by Slapbox** [comment]
To elaborate, we need to define the window `icon` for the sake of KDE environments at the very least. Without defining one, both the taskbar and the task switcher will simply show an 'X'.

On Windows it falls back to the application icon, but not so with KDE.

**Comment by Slapbox** [comment]
@nornagon see the updated repro which demonstrates the issue without any reliance on `global` -- https://github.com/Slapbox/electron-icon-memory-usage-bug

**Comment by nornagon** [comment]
NativeWindowViews stores a copy (refcounted?) of the HICON: https://github.com/electron/electron/blob/e0f2511cba757925ade70a3936fde14339a2504e/shell/browser/native_window_views.cc#L1454-L1456

And I'm not actually sure it uses scales other than 1x? Here are the steps:

https://github.com/electron/electron/blob/e0f2511cba757925ade70a3936fde14339a2504e/shell/browser/api/electron_api_base_window.cc#L102-L103

calls `NativeImage::CreateFromPath`, which loads all the representations, then:

https://github.com/electron/electron/blob/e0f2511cba757925ade70a3936fde14339a2504e/shell/browser/api/electron_api_base_window.cc#L1058-L1061

If you're using PNGs and not ICOs, `GetHICON` creates an icon using `image_.AsBitmap()`:

https://github.com/electron/electron/blob/e0f2511cba757925ade70a3936fde14339a2504e/shell/common/api/electron_api_native_image.cc#L196-L197

`image_.AsBitmap()` grabs the `gfx::ImageSkia` and calls `->bitmap()` on it:

https://source.chromium.org/chromium/chromium/src/+/main:ui/gfx/image/image.cc;l=410-413;drc=e84b72033fb15d2827fee6f2cb2b1e7fd0992c25
```cpp
const SkBitmap* Image::ToSkBitmap() const {
  // Possibly create and cache an intermediate ImageRepSkia.
  return ToImageSkia()->bitmap();
}
```

`::bitmap()` calls `GetBitmap()` which calls `FindRepresentation(1.0)`, i.e. it requests the 1x representation:

https://source.chromium.org/chromium/chromium/src/+/main:ui/gfx/image/image_skia.cc;l=535;drc=e84b72033fb15d2827fee6f2cb2b1e7fd0992c25
```cpp
  auto it = storage_->FindRepresentation(1.0f, true);
```


It looks like after all that, the image gets discarded and only the HICONs are stored, so by this analysis I'd expect roughly as much memory to be used as it takes to store the 1x bitmap, which would be ~4 MB for a 1024x1024 icon.


... and by the way, why is your icon so big?? 1024x1024 is _massive_!

**Comment by Slapbox** [comment]
I agree, the icon is way too big!

**TLDR:** It's temporary and the size we'll be using in the end is likely to be 256x256

**Long version:** Basically I wasn't sure what size would be the smallest necessary to fill the KDE largest task switcher icons without any stretching, and for LatteDock also. So I started out with a very large image to be sure we'd fill it, figuring that I'll release a patch in a couple weeks to bring it down to the ideal size (not realizing at that point that the icons were stored in memory as bitmaps) - _but_ when I found that the high DPI icons increased memory usage so dramatically I decided to include only a 1024x1024 image, since even in high DPI scenarios it should look decent.

Additionally, I didn't want to change it while in the middle of discussion/investigation into what was causing the greatly increased memory usage.

----

But it really does seem like the memory is not being released for high DPI icons in that repro, and in our app. The development environment v8 heap snapshot size in our app is 309mb with or without high DPI icons included - but memory usage with high DPI icons is ~300mb higher in the packaged app, and 834mb higher in development (2256mb vs 1422mb.)

Even dropping the size down to 256x256, the expected memory usage would be nearly 20mb if the unused variants aren't removed (300mb / 4<sup>2</sup>,) as opposed to an expected 0.25mb for standard DPI screens. That's still rather considerable on low resource machines.


----

Regarding:

> And I'm not actually sure it uses scales other than 1x?

Makes sense! In most environments I think this is the most logical choice - but in KDE the window icon affects the app switcher icon, and I believe [LatteDock](https://github.com/KDE/latte-dock) as well - and the icons can become quite large in LatteDock and look quite pixelated without sufficiently high resolution image. I'm not aware of any similar necessity on Windows/macOS.

**Comment by nornagon** [comment]
FWIW the above investigation was limited to the Windows code path, I didn't look into it on Linux because of your earlier suggestion that the problem wasn't showing up on Linux, but looking back I see you did end up reproducing it on Linux. That said, it seems like the Linux (X11) version _also_ just fetches the 1x representation: https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/platform/x11/x11_window.cc;l=961-983;drc=1a847f7d83d6280fb0e8965fbd3e1077c18745fc

**Comment by Slapbox** [comment]
Hm, that's so strange that the memory usage increases so much even though only the 1x representations are actually fetched for presentation in the UI! I won't pretend I fully grok all the linked code, though I at least follow the `window_icon.GetRepresentation(1.0f)`.

With that in mind, I guess it currently makes no sense for us to include high DPI icons for Window/Tray icons? But these seem like the main things I'd really expect high DPI variations to exist for; as SVGs are available, and often preferable, in `BrowserWindows`. 

----

But besides the fact that including variations for these icons is currently not useful, it still seems like something is wrong with the method of removing unused data from `nativeImage` objects. I also won't pretend to understand how C++ garbage collection works, or if it's even automatic in this circumstance, but it's really counterintuitive that all variations/representations end up in memory, and remain there.

I wonder if anything can be gleaned from the fact that, for the memory usage increase to occur, the `nativeImage` must be assigned to a `BrowserWindow` icon. It doesn't seem like the memory can be accounted for in `v8` (memory allocation is the same whether variations are included/not included/no icon is defined), so I wonder if that means that the memory is not being deallocated on the native side of things, and if so, is that a design limitation of `nativeImage` as currently implemented?

**Comment by nornagon** [comment]
Hmm, I don't seem to be able to reproduce that result on Linux.

```js
const {app, nativeImage, BrowserWindow} = require('electron')
const path = require('path')
setInterval(() => {
  gc()
  console.log(app.getAppMetrics().find(x => x.type === 'Browser').memory)
}, 1000)

app.whenReady().then(() => {
  const iconPath = path.resolve('build/icons/icon-1024x1024.png');

  new BrowserWindow({
    // Commenting out the line below doesn't significantly change the app's reported memory usage.
    icon: nativeImage.createFromPath(iconPath)
  })
})
```

Both in `htop` and in the memory numbers reported by `app.getAppMetrics()`, I see a memory footprint of about 140 MB with or without the icon specified.

**Comment by nornagon** [comment]
I bisected to find when the behavior changed, and it looks like this was fixed in the range [v14.0.0-nightly.20210413...v14.0.0-nightly.20210426](https://github.com/electron/electron/compare/v14.0.0-nightly.20210413...v14.0.0-nightly.20210426)

**Comment by nornagon** [comment]
I suspect the fix was https://github.com/electron/electron/commit/4ca518468d015a2d78524cad88e9917775921232#diff-bfaa73008a8b883d15dd11323996e12e24ae556c8a03e8dd0ad43b8785cf099d, it looks like probably we were previously retaining a reference to the icon in order to pass it on to child windows, but with that change we ceased to do so, allowing the icon to be garbage-collected.

**Comment by nornagon** [comment]
Going to close this as I think the issue has been adequately addressed.

**Comment by Slapbox** [comment]
Confirming everything looks good with Electron@17.0.0! Thank you so much for your help with this @nornagon.

----

I'm still confused about why the DPI variations don't apply to things like the tray icon though?

On Windows 11 the tray icon size is 16x16px. If the system scaling is set to 125% then the tray icon size becomes 20x20px, and the 16x16px icon is upscaled (poorly) by the OS. On the other hand, using an icon that's too large is subject to being downscaled (also poorly) by the OS. These things weren't issues on Windows 7, not that I could see anyway, so I'm guessing the algorithms for image scaling have been changed.

It would be a big step up in terms of polish if DPI variations were used in these cases.

# DPI 125%:

**16x16px icon**
![image](https://user-images.githubusercontent.com/6835891/153320834-54127645-ecb9-4084-8455-6e0fa54ad646.png)

**1024x1024px icon**
![image](https://user-images.githubusercontent.com/6835891/153320876-4342ca67-8978-4c05-a82a-e60fcd800931.png)

# DPI 100%

**16x16px icon**
![image](https://user-images.githubusercontent.com/6835891/153320930-7ae6c38c-3767-4e68-944c-4306c00941b4.png)

**1024x1024px icon**
![image](https://user-images.githubusercontent.com/6835891/153320956-19f40874-8c71-4afc-912a-c14ac066a7df.png)

----

Providing any size besides 16px for downscaling results in varying bad outcomes. Even 32x32px looked bad in my testing, so simply providing moderately oversized images doesn't seem a viable workaround.

**Comment by Slapbox** [comment]
Sorry to be a pain, I remain confused about whether there's any resources for how this sort of stuff should be handled, or if it can be handled? 

Do developers just need to wait for further development in Electron itself to support these cases? And if so, are there any plans by the Electron Team for handling those cases?

It seems impossible to properly display app icons on high DPI screens like Retina displays and like the only potential workaround is to provide oversized images (which doesn't work well in Windows as shown above.)

**Comment by nornagon** [comment]
Yeah, it seems like Electron currently only supports a single size of tray icon:

https://github.com/electron/electron/blob/83a4ac184117c9b60732ba893be0d1212d9231d0/shell/browser/api/electron_api_tray.cc#L199-L203

Please open a new issue to track fixing that!

**Comment by Slapbox** [comment]
Many thanks for all your help @nornagon! New issue created at #33044.
